Add `MelSpectrogram` layer

awsaf49 opened the pull request:
This PR will add the `MelSpecrtoram` layer as an `audio_preprocessing` layer as mentioned in #17657. I have added the backbone of this layer. This layer will convert raw audio signals to Mel spectrograms. This layer is compatible with both GPU & TPU.

Need to add some tests to ensure everything is okay.

## Todo
- [x] unbatched audio test
- [x] batched audio test
- [x] zero values audio test
- [ ] serialize callable `ref`

cc: @fchollet , @mattdangerw

awsaf49 commented:
It turns out linting is failing due to conflict between `flake8` and `black`. Black is making the formatted code line too wide which is flagged by flake8 on other hand if code is reformated  to adjust for too wide line then black flags it

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+
+
+@keras_export(
+    "keras.layers.MelSpectrogram",
```
You can remove this line -- we'll export it to experimental for now.

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+
+@keras_export(
+    "keras.layers.MelSpectrogram",
+    "keras.layers.experimental.preprocessing.MelSpectrogram",
```
keras.layers.experimental.MelSpectrogram

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+class MelSpectrogram(base_layer.Layer):
+    """A preprocessing layer to convert raw audio signals to Mel spectrograms.
+
+    This layer takes `float32/float64` single or batched audio signal as inputs
```
`float32`/`flaot64`

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+    representing audio signals. The output will be a 2D or 3D tensor
+    representing Mel spectrograms.
+
+    For an overview and full list of preprocessing layers, see the preprocessing
```
Rather than a preprocessing layers reference it would be useful to include info about Mel spectrograms, what they are, why use them, etc. A link to a page would be nice

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+        1D (unbatched) or 2D (batched) tensor with shape:`(..., samples)`.
+
+    Output shape:
+        2D (unbatched) or 3D (batched) tensor with shape:`(..., n_mels, time)`.
```
Use `num_` for numbers rather than `n_`

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+        2D (unbatched) or 3D (batched) tensor with shape:`(..., n_mels, time)`.
+
+    Args:
+        n_fft: Integer, size of the FFT window.
```
fft_window_size?

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+            1 for magnitude, 2 for power, etc. Default is 2.
+        ref: Float, the power is scaled relative to it `10 * log10(S / ref)`.
+        amin: Float, minimum value for power and `ref`.
+    """
```
Please add a simple code example in the docstring, for the most common use case.

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+        return config
+
+
+def convert_inputs(inputs, dtype=None):
```
This is not needed, just inline the function where it is used

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+        top_db=80.0,
+        power=2.0,
+        amin=1e-10,
+        ref=1.0,
```
We need to figure out the best way to name these arguments. Some rules:

- Always prefer fully-spelled out names over abbreviations, unless the abbreviation is completely obvious
- Someone already familiar with what Mel spectrograms are should understand the meaning of all arguments just from their names
- We should be consistent with naming conventions across the Keras API, for instance this API (and others) https://keras.io/api/preprocessing/timeseries/

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+        win_length=None,
+        window="hann_window",
+        sr=16000,
+        n_mels=128,
```
`num_mel_bins`

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+        hop_length=512,
+        win_length=None,
+        window="hann_window",
+        sr=16000,
```
`sampling_rate`

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+        self,
+        n_fft=2048,
+        hop_length=512,
+        win_length=None,
```
Probably `fft_frame_length` (not sure)

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+    def __init__(
+        self,
+        n_fft=2048,
+        hop_length=512,
```
Probably `fft_stride`

fchollet commented on keras/layers/preprocessing/audio_preprocessing.py:
```diff
@@ -0,0 +1,194 @@
+
+    def __init__(
+        self,
+        n_fft=2048,
```
Not sure what to call this, perhaps `fft_size`. Or could it be `num_fft_bins`?

fchollet reviewed (commented):
Thanks for the PR! This is good work. There's a lot of API design work to be done here to make the API as intuitive as possible.

awsaf49 commented:
@fchollet Thanks for your valuable feedback. While I understand your concern about the potential confusion caused by non-intuitive names, I would like to point out that the current argument names are widely used and recognized in the community, as demonstrated by popular libraries such as `librosa`,`torchaudio`, and `nnaudio`. For example,

for `librosa`
```py
librosa.feature.melspectrogram(y=None, sr=22050, n_fft=2048, n_mels=128, hop_length=512, win_length=None, 
                               window='hann', power=2.0, fmin=0.0, fmax=None)
```
for `torchaudio`
```py
torchaudio.transforms.MelSpectrogram(sample_rate=16000, n_fft=400, win_length=None, hop_length=None, 
                                     f_min=0.0, f_max= None, n_mels = 128, window_fn = <fn>, power = 2.0)
```
for `nnaudio`,
```py
nnAudio.Spectrogram.MelSpectrogram(sr=22050, n_fft=2048, n_mels=128, hop_length=512, 
                                   window='hann', power=2.0,fmin=0.0, fmax=None)
```
While I am open to alternative names, I believe that changing them could create confusion for users who are already familiar with these naming conventions. Therefore, it seems like a trade-off between renowned names and intuitive names. Any thoughts what should we do?

fchollet commented:
> Therefore, it seems like a trade-off between renowned names and more intuitive names. Any thoughts what should we do?

We should use more intuitive names.

- There are more people who will use these APIs in the future than they are people using these APIs today. We're doing them a service by adopting better naming conventions.
- If the names are intuitive, then they will be intuitive / easy to understand for people already familiar with the current APIs.
- Keras APIs must be consistent with the Keras API. It would be annoying and surprising if something called "sampling_rate" in several other places of the API was called "sample_rate" here.

awsaf49 commented:
@fchollet I've updated the names according to your suggestions and replaced the rest with better intuitive names. Let me know if they meet the requirements.

 By the way just noticed, this PR keras-team/keras-nlp/pull/847 in Keras-NLP is using `sample_rate` instead of `sampling_rate` & `stride` instead of `fft_stride`.

fchollet reviewed (commented):
Thanks for the update -- the API is looking good (just one comment)! Please add unit tests.

awsaf49 commented:
@fchollet could you please check?

awsaf49 commented:
@gbaned any update?

gbaned commented:
> @gbaned any update?

Hi @awsaf49  Sorry for the delay, this PR is in awaiting review status.

awsaf49 commented:
This pull request has been on hold for quite some time. @fchollet and @mattdangerw, would you kindly consider reviewing it? I strongly believe that this new layer will be helpful to people working with audio data. Thus, the sooner it is added into Keras, the sooner the community can get its benefits.

awsaf49 commented:
@gbaned could you approve the workflow for **unit-test** ??

awsaf49 commented:
@mihirparadkar Hi, I just noticed `keras-team-review-pending` label has been removed. Any update on this??

fchollet reviewed (approved):
LGTM

awsaf49 commented:
Currently the PR is on hold due to the following error,
```shell
/keras/distribute/ctl_correctness_test.runfiles/org_keras/keras/metrics/confusion_metrics.py", line 22, in <module>
    from keras import activations
  File "/home/kbuilder/.cache/bazel/_bazel_kbuilder/31d6f47147b75c35404d734345be7323/execroot/org_keras/bazel-out/k8-opt/bin/keras/distribute/ctl_correctness_test.runfiles/org_keras/keras/activations.py", line 22, in <module>
    import keras.layers.activation as activation_layers
ImportError: cannot import name 'layers' from partially initialized module 'keras' (most likely due to a circular import) (/home/kbuilder/.cache/bazel/_bazel_kbuilder/31d6f47147b75c35404d734345be7323/execroot/org_keras/bazel-out/k8-opt/bin/keras/distribute/ctl_correctness_test.runfiles/org_keras/keras/__init__.py)
```
I can't seem to find a fix for it as it relates to `keras circular import` but the code is written with a similar structure as the `image_processing` layer which works just fine. Any suggestions would be really helpful.
@fchollet @mattdangerw

sachinprasadhs commented:
Hello, Thank you for submitting a pull request.

We're currently in the process of migrating the new `Keras 3 code base` from `keras-team/keras-core` to `keras-team/keras`. 
Consequently, merging this PR is not possible at the moment. After the migration is successfully completed, feel free to reopen this PR at `keras-team/keras` if you believe it remains relevant to the `Keras 3 code base`. If instead this PR fixes a bug or security issue in legacy `tf.keras`, you can instead reopen the PR at `keras-team/tf-keras`, which hosts the `TensorFlow-only`, legacy version of `Keras`.
